security/pkg/pki/util/generate_cert_test.go

```diff
@@ -551,3 +551,71 @@ func TestAppendRootCerts(t *testing.T) {
 		}
 	}
 }
+
+func TestGetCertOptionsFromExistingCert(t *testing.T) {
+	caCertTTL := 24 * time.Hour
+	oldOrg := "old org"
+	caKeySize := 512
+
+	caCertOptions := CertOptions{
+		TTL:          caCertTTL,
+		Org:          oldOrg,
+		IsCA:         true,
+		IsSelfSigned: true,
+		RSAKeySize:   caKeySize,
```

Comment:
Can we also test with a different caKeySize?
See it's configurable so we need to consider the case that this is changed: https://github.com/istio/istio/blob/master/security/cmd/istio_ca/main.go#L233

Currently we don't get the RSAKeySize in GetCertOptionsFromExistingCert. I doubt whether we need get that to ensure the new cert is using the same key size. But we may not need it because we are explicitly setting the public and private keys when creating the cert.

Either way, we need to support this case and comment what values are preserved from the old cert.

PS: I also noticed the dualUse is configurable. But I don't think it's actually used in cert generation. Could you please double check?

Reply:
The RSAKeySize does not take any effect when we generate root certificate from an existing private key. The key size is get from the existing private key. I have added a check for the key size.
Here are more details: 

1. The original root cert and key is generated by calling [GenCertKeyFromOptions()](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/util/generate_cert.go#L87). RSAKeySize is passed to rsa.GenerateKey(), which returns *rsa.PrivateKey. 
2.  When generating new root cert with old private key, we call [GenRootCertFromExistingKey()](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/util/generate_cert.go#L132), and inside that function, the old private key PEM is passed to ParsePemEncodedKey(), which returns *rsa.PrivateKey. That stores the old RSAKeySize information.

Currently, IsDualUse is not used. When the CA generates the original root certificate, the cert options look like [this](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/ca/ca.go#L138). IsDualUse is passed from configuration but ```Host``` is not set. The CA then calls [GenCertKeyFromOptions()](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/util/generate_cert.go#L87), which calls [genCertTemplateFromOptions()](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/util/generate_cert.go#L251). Inside genCertTemplateFromOptions() only Host is non-empty, and IsDualUse is true, the the template sets CommonName. You can check [here](https://github.com/istio/istio/blob/94c8d78639bcc92502e0ecb148abea51eb94334d/security/pkg/pki/util/generate_cert.go#L284). For now, I think we don't need to worry about extract Host and IsDualUse information from old root certificate, because they are not used. But in the future we need to cover that part, and it is not trivial.

Reply:
Thanks for adding the tests.
I think the DualUse doesn't matter for the root certs. It's a configuration for generating workload certs.